Allow retrieval of project layer order through QgsProject

nyalldawson opened the pull request:
Previously this was only accessible through app.

Additionally, make sure that map themes always respect project layer order and do not store out of date layer orders, resulting in a different rendered order between composer and canvas.

nyalldawson committed:
Allow retrieval of project layer order through QgsProject

Previously this was only accessible through app

nyalldawson committed:
Layer order in map themes must always respect project layer order

nyalldawson committed:
Project layer order returns all layers in order, not just visible ones

Otherwise map themes with different visible layers cannot be
correctly ordered

m-kuhn commented on src/core/qgsproject.h:
```diff
@@ -951,6 +975,8 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
 
     QMap<QString, QgsMapLayer *> mMapLayers;
 
+    QgsWeakMapLayerPointerList mLayerOrder;
```
Can that be moved to `QgsMapThemeCollection`?

nyalldawson replied:
Does it really makes any difference? Happy to, but @wonder-sk preferred it here.

m-kuhn replied:
While we are trying to move away from QgsProject as singleton - and that's a good move! - it will still keep being something similar, as in there is always one QgsProject per workspace/project. So everything that is in QgsProject will be available only once per workspace.

As such it comes with very similar properties as a singleton: it's very easy to implement but doesn't allow for flexibility. Therefore I'd like to reduce the number of direct members of QgsProject to a minimum and am looking for some other place to put this.

Apart from this (in practice we also only have one QgsMapThemeCollection per QgsProject) I would like to avoid the QgsProject to become a "general dumping ground" for everything.

nyalldawson committed:
If no project layer order is set, fall back to stored layer order
when retrieving map theme layers

Since we don't ever want to return no layers in this situation
using the stored order is an acceptable last resort

Should only affect custom scripts run outside of QGIS app
in any case!

nyalldawson committed:
Auto-set GUI for custom layer order when API call to QgsProject::setLayerOrder
is made

nyalldawson committed:
No longer need to store custom layer order for canvas bridge

Since it's now set at the project level

nyalldawson committed:
Qt5 style connects

nyalldawson committed:
Add method to QgsMapThemeCollection to retrieve ordered list of
visible layers for the project

nyalldawson committed:
Remove no longer required hack to get composer maps to render
canvas layers

nyalldawson committed:
Indentation

nyalldawson committed:
Move test to correct location

nyalldawson committed:
Clearing map canvas theme falls back to project layers

nyalldawson committed:
If no project layer order set, use layer tree layer order
for QgsMapThemeCollection::masterVisibleLayers()

nyalldawson committed:
When theme is removed, canvas falls back to default view

nyalldawson committed:
Fix failing composer map tests

nyalldawson committed:
Revert "Remove no longer required hack to get composer maps to render"

Because... server

3nids commented:
thanks for tackling this!
is this ready to merge ?

nyalldawson commented:
I think so. It's been heavily reviewed today by @wonder-sk (via hangouts).

3nids commented:
thanks!
